src/pgwire/src/protocol.rs

```diff
@@ -407,20 +425,36 @@ where
     }
 
     async fn advance_drain(&mut self) -> Result<State, io::Error> {
-        match self.conn.recv().await? {
+        let message = self.conn.recv().await?;
+        if message.is_some() {
+            self.adapter_client()
+                .remove_idle_in_transaction_session_timeout();
+        }
+        match message {
             Some(FrontendMessage::Sync) => self.sync().await,
             None => Ok(State::Done),
             _ => Ok(State::Drain),
         }
     }
 
+    async fn advance_timed_out(&mut self, err: ErrorResponse) -> Result<State, io::Error> {
+        // dropping the adapter client will terminate the session.
+        self.adapter_client = None;
```

Comment:
Why do we need to do it exactly this way? Could this revert the `Option<Client>` change here and instead this function explicitly sends a `Terminate` message? That would perhaps enable us to not add the `TimedOut` variant which I'm uneasy about adding, because we could do the logic right where we get the timeout message, and return `Done` there.

Reply:
I just pushed a new commit that moved this logic inline so we can get rid of the `TimedOut` state.

The reason for `Option<Client>` is the following:
1. Terminating a session consumes the `Session` object.
2. The `AdapterClient` object has an invariant that `Session` is always `Some` after any method returns.

I was trying to implement this without changing either of those two things to try and leave a smaller footprint. However I just found a bug while removing `TimedOut` that may require me to change this. Sending an error message back to the client requires access to session variables to see the log level. I'm not sure how that wasn't caught in the tests in CI, because locally the test panics for me.